module/builder/collection/builder.go

```diff
@@ -345,3 +306,88 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 
 	return proposal.Header, err
 }
+
+// populateUnfinalizedAncestryLookup traverses the unfinalized ancestry backward
+// to populate the transaction lookup (used for deduplication) and the rate limiter
+// (used to limit transaction submission by payer).
+//
+// The traversal begins with the block specified by parentID (the block we are
+// building on top of) and ends with the oldest unfinalized block in the ancestry.
+func (b *Builder) populateUnfinalizedAncestryLookup(parentID flow.Identifier, finalHeight uint64, lookup *transactionLookup, limiter *rateLimiter) error {
+
+	ancestorID := parentID
+	for {
+		ancestor, err := b.clusterHeaders.ByBlockID(ancestorID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor header: %w", err)
+		}
+
+		if ancestor.Height <= finalHeight {
+			break
+		}
+
+		payload, err := b.payloads.ByBlockID(ancestorID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor payload: %w", err)
+		}
+
+		for _, tx := range payload.Collection.Transactions {
+			lookup.addUnfinalizedAncestor(tx.ID())
+			limiter.addAncestor(ancestor.Height, tx)
+		}
+		ancestorID = ancestor.ParentID
+	}
+
+	return nil
+}
+
+// populateFinalizedAncestryLookup traverses the reference block height index to
+// populate the transaction lookup (used for deduplication) and the rate limiter
+// (used to limit transaction submission by payer).
+//
+// The traversal is structured so that we check every collection whose reference
+// block height translates to a possible constituent transaction which could also
+// appear in the collection we are building.
+func (b *Builder) populateFinalizedAncestryLookup(minRefHeight, maxRefHeight uint64, lookup *transactionLookup, limiter *rateLimiter) error {
```

Comment:
We have unified functions for fork traversal here:
* https://github.com/onflow/flow-go/blob/6aba4c0f95d4d7e2a71d8c34ff1ec2829f285aec/state/fork/traversal.go#L16
* https://github.com/onflow/flow-go/blob/6aba4c0f95d4d7e2a71d8c34ff1ec2829f285aec/state/fork/traversal.go#L45

They are designed to be highly customizable. Usage examples:
* https://github.com/onflow/flow-go/blob/6aba4c0f95d4d7e2a71d8c34ff1ec2829f285aec/module/builder/consensus/builder.go#L305
* https://github.com/onflow/flow-go/blob/6aba4c0f95d4d7e2a71d8c34ff1ec2829f285aec/module/validation/seal_validator.go#L191 

Not sure, but hopefully we are able to use this here and avoid replicating this logic in many different places over and over.

Reply:
👍  Good point - I refactored the un-finalized ancestry lookup to use the fork traversal functions. I don't think it's applicable to finalized ancestry though, since we're traversing using the reference height index instead of by block parent.

Reply:
If we just traverse the fork for finalized blocks, we don't know when to stop the traversal. It's also possible we might unnecessarily check collections that can't possibly contain a duplicate transaction. That's the reason for adding the index in this PR. The index tells us, given a collection's reference block height, exactly which other collections can possibly contain a conflicting transaction. 

We need to check for duplicates in any collection with an expiry range that overlaps with the candidate collection's expiry range. Since expiry is defined by heights on the main chain, and we're traversing the cluster chain, we can't just traverse backward a fixed number of collections. We also can't assume that, if we observe a collection with a reference height outside the range overlapping with the candidate collection, it means that no other older ancestor will have a reference height that does overlap with the candidate collection. Consecutive collections may not have strictly increasing reference block heights.

Reply:
Thank you for explanation, yeah, that makes sense